Non-deterministic test failures

We sometimes observe a failing test:
```
=== RUN   TestHandlerDispatcherMonitoring
time="2019-01-21T14:21:21Z" level=info msg="submission to tdh.flows () successful" domain=submitter payloadsize=0 rawsize=0 submitter=AMQP
time="2019-01-21T14:21:23Z" level=info msg="submission to tdh.flows () successful" domain=submitter payloadsize=0 rawsize=0 submitter=AMQP
time="2019-01-21T14:21:25Z" level=info msg="submission to tdh.flows () successful" domain=submitter payloadsize=0 rawsize=0 submitter=AMQP
time="2019-01-21T14:21:27Z" level=info msg="submission to tdh.flows () successful" domain=submitter payloadsize=0 rawsize=0 submitter=AMQP
time="2019-01-21T14:21:29Z" level=info msg="submission to tdh.flows () successful" domain=submitter payloadsize=0 rawsize=0 submitter=AMQP
time="2019-01-21T14:21:30Z" level=info msg="submission to nsm.test.metrics () successful" domain=submitter payloadsize=69 rawsize=69 submitter=AMQP
time="2019-01-21T14:21:30Z" level=info msg="closing connection" domain=submitter submitter=AMQP
--- FAIL: TestHandlerDispatcherMonitoring (10.00s)
	handler_dispatcher_test.go:233: unexpected result length: 0 == 0
```

--- a/processing/forward_handler.go
+++ b/processing/forward_handler.go
@@ -37,6 +37,8 @@ type ForwardHandler struct {
 	StatsEncoder        *util.PerformanceStatsEncoder
 	StopChan            chan bool
 	StoppedChan         chan bool
+	StopCounterChan     chan bool
+	StoppedCounterChan  chan bool
 	Running             bool
 	Lock                sync.Mutex
 }
@@ -136,21 +138,23 @@ func (fh *ForwardHandler) runForward() {
 }
 
 func (fh *ForwardHandler) runCounter() {
-	var nofSecs uint64 = 10
+	sTime := time.Now()
 	for {
+		time.Sleep(500 * time.Millisecond)
 		select {
-		case <-fh.StopChan:
+		case <-fh.StopCounterChan:
+			close(fh.StoppedCounterChan)
 			return
 		default:
-			time.Sleep(time.Duration(nofSecs) * time.Second)
-			fh.Lock.Lock()
-			if fh.StatsEncoder != nil {
-				fh.PerfStats.ForwardedPerSec /= nofSecs
-				fh.StatsEncoder.Submit(fh.PerfStats)
+			if fh.StatsEncoder == nil || time.Since(sTime) < fh.StatsEncoder.SubmitPeriod {
+				continue
 			}
+			fh.Lock.Lock()
+			fh.PerfStats.ForwardedPerSec /= uint64(fh.StatsEncoder.SubmitPeriod.Seconds())
+			fh.StatsEncoder.Submit(fh.PerfStats)
 			fh.PerfStats.ForwardedPerSec = 0
+			sTime = time.Now()
 			fh.Lock.Unlock()
-
 		}
 	}
 }
@@ -202,6 +206,8 @@ func (fh *ForwardHandler) Run() {
 	if !fh.Running {
 		fh.StopChan = make(chan bool)
 		fh.ForwardEventChan = make(chan []byte, 10000)
+		fh.StopCounterChan = make(chan bool)
+		fh.StoppedCounterChan = make(chan bool)
 		go fh.reconnectForward()
 		fh.ReconnectNotifyChan <- true
 		go fh.runForward()
@@ -213,6 +219,8 @@ func (fh *ForwardHandler) Run() {
 // Stop stops forwarding of JSON representations of all consumed events
 func (fh *ForwardHandler) Stop(stoppedChan chan bool) {
 	if fh.Running {
+		close(fh.StopCounterChan)
+		<-fh.StoppedCounterChan
 		fh.StoppedChan = stoppedChan
 		fh.Lock.Lock()
 		fh.OutputConn.Close()

--- a/processing/forward_handler_test.go
+++ b/processing/forward_handler_test.go
@@ -134,7 +134,7 @@ func TestForwardHandler(t *testing.T) {
 	<-cldCh
 
 	if len(coll) != 2 {
-		t.Fatal("unexpected number of alerts")
+		t.Fatalf("unexpected number of alerts: %d", len(coll))
 	}
 	if coll[0] != "foo1" {
 		t.Fatalf("invalid payload, expected 'foo1', got %s", coll[0])

--- a/processing/handler_dispatcher.go
+++ b/processing/handler_dispatcher.go
@@ -58,20 +58,22 @@ func (h *DefaultHandler) Consume(e *types.Entry) error {
 }
 
 func (ad *HandlerDispatcher) runCounter() {
-	var nofSecs uint64 = 10
+	sTime := time.Now()
 	for {
+		time.Sleep(500 * time.Millisecond)
 		select {
 		case <-ad.StopCounterChan:
 			close(ad.StoppedCounterChan)
 			return
 		default:
-			time.Sleep(time.Duration(nofSecs) * time.Second)
-			ad.Lock.Lock()
-			if ad.StatsEncoder != nil {
-				ad.PerfStats.DispatchedPerSec /= nofSecs
-				ad.StatsEncoder.Submit(ad.PerfStats)
+			if ad.StatsEncoder == nil || time.Since(sTime) < ad.StatsEncoder.SubmitPeriod {
+				continue
 			}
+			ad.Lock.Lock()
+			ad.PerfStats.DispatchedPerSec /= uint64(ad.StatsEncoder.SubmitPeriod.Seconds())
+			ad.StatsEncoder.Submit(ad.PerfStats)
 			ad.PerfStats.DispatchedPerSec = 0
+			sTime = time.Now()
 			ad.Lock.Unlock()
 		}
 	}

--- a/processing/handler_dispatcher_test.go
+++ b/processing/handler_dispatcher_test.go
@@ -12,6 +12,7 @@ import (
 
 	"github.com/DCSO/fever/types"
 	"github.com/DCSO/fever/util"
+	log "github.com/sirupsen/logrus"
 
 	"github.com/NeowayLabs/wabbit"
 	"github.com/NeowayLabs/wabbit/amqptest"
@@ -173,6 +174,7 @@ func TestHandlerDispatcherMonitoring(t *testing.T) {
 	results := make([]string, 0)
 	c, err := util.NewConsumer(serverURL, "nsm.test.metrics", "direct", "nsm.test.metrics.testqueue",
 		"", "", func(d wabbit.Delivery) {
+			log.Info(string(d.Body()))
 			results = append(results, string(d.Body()))
 		})
 	if err != nil {
@@ -200,28 +202,36 @@ func TestHandlerDispatcherMonitoring(t *testing.T) {
 	ad := MakeHandlerDispatcher(outChan)
 	ad.SubmitStats(pse)
 	ad.Run()
-	time.Sleep(5 * time.Second)
+
+	go func() {
+		for i := 0; i < 100; i++ {
+			ad.Dispatch(&types.Entry{
+				JSONLine: "foo",
+			})
+			ad.Dispatch(&types.Entry{
+				JSONLine: "bar",
+			})
+			ad.Dispatch(&types.Entry{
+				JSONLine: "baz",
+			})
+			time.Sleep(50 * time.Millisecond)
+		}
+	}()
 
 	go func(closeChan chan bool, inChan chan types.Entry) {
+		i := 0
 		for v := range inChan {
 			_ = v
+			i++
+			if i == 300 {
+				break
+			}
 		}
 		close(closeChan)
 	}(closeChan, outChan)
 
-	ad.Dispatch(&types.Entry{
-		JSONLine: "foo",
-	})
-	ad.Dispatch(&types.Entry{
-		JSONLine: "bar",
-	})
-	time.Sleep(3 * time.Second)
-	ad.Dispatch(&types.Entry{
-		JSONLine: "baz",
-	})
-
-	close(outChan)
 	<-closeChan
+	close(outChan)
 
 	stopChan := make(chan bool)
 	ad.Stop(stopChan)
@@ -230,9 +240,10 @@ func TestHandlerDispatcherMonitoring(t *testing.T) {
 	c.Shutdown()
 
 	if len(results) == 0 {
-		t.Fatalf("unexpected result length: %d == 0", len(results))
+		t.Fatalf("unexpected result length: 0")
 	}
-	if match, _ := regexp.Match(fmt.Sprintf("^%s,[^ ]+ dispatch_calls_per_sec=[0-2]", util.ToolName), []byte(results[0])); !match {
+
+	if match, _ := regexp.Match(fmt.Sprintf("^%s,[^ ]+ dispatch_calls_per_sec=[0-9]+", util.ToolName), []byte(results[0])); !match {
 		t.Fatalf("unexpected match content: %s", results[0])
 	}
 }

